### PR TITLE
some minor improvement

### PR DESCRIPTION
**Pull request by alphatownsman**
- added an option to allow login via any mastodon instance (schema migration required)
- added a few options to make it slightly easier to customize and install as new site
- improve compatibility w Pleroma (schema migration required)
- added an option to customize permission scope when connecting to Mastodon

these options have default values to keep current behaviors as is, although they might not be optimal for a new site

### CODE REVIEW

File: mastodon/api.py [hunk 1]
```diff
@@ -135,7 +137,11 @@ def get_relationship(request_user, target_user, token):
         return get_relationships(request_user.mastodon_site, target_user.mastodon_id, token)
     else:
         cross_site_id = get_cross_site_id(target_user, request_user.mastodon_site, token)
-        return get_relationships(request_user.mastodon_site, [cross_site_id,], token)
+        if cross_site_id is None:
+            return [{'blocked_by': True}]  # boldly assume blocked(?!) if no relationship found
+            # FIXME should check the reverse direction? but need either cache the target user's oauth token or her blocked list
```
Comment (alphatownsman) [inline mastodon/api.py:142]: Choose to be strict here but not sure if this is best option. Thoughts?

File: mastodon/api.py [hunk 2]
```diff
@@ -121,9 +119,13 @@ def get_site_id(username, user_site, target_site, token):
...
     }
-    response = get(url, data=payload, headers=headers)
+    response = get(url, params=payload, headers=headers)
     data = response.json()
-    if not data['accounts']:
+    if not data['accounts']: 
+        return None
+    elif len(data['accounts']) == 0:  # target site may return empty if no cache of this user
+        return None
+    elif data['accounts'][0]['acct'] != f"{username}@{user_site}":  # or return another user with a similar id which needs to be skipped
```
Comment (alphatownsman) [inline mastodon/api.py:128]: this is tricky and not ideal too.

**Review by alphatownsman — COMMENTED** [review]
@doubaniux please see some comments/questions inline, and let me know if you are cool with merge this. Thanks 🙏

**Comment by joseph-cheung** [comment]
I'm trying to run your fork, but failed.
Wondering what "sync_synctask" is. While `makemigrations`, it errors out with this message:
```
django.db.utils.ProgrammingError: relation "sync_synctask" does not exist
LINE 1: ..."started_time", "sync_synctask"."ended_time" FROM "sync_sync...
```

**Comment by alphatownsman** [comment]
@joseph-cheung there's a comment above :
https://github.com/doubaniux/boofilsic/pull/92#discussion_r708789055

just uncomment that line and you should be fine

the reason I did not uncomment in the MR is bc I don't know why it's missed in the first place.

**Comment by joseph-cheung** [comment]
> @joseph-cheung there's a comment above :
> [#92 (comment)](https://github.com/doubaniux/boofilsic/pull/92#discussion_r708789055)
> 
> just uncomment that line and you should be fine
> 
> the reason I did not uncomment in the MR is bc I don't know why it's missed in the first place.

Sorry but I did uncommented that line as you said above... Maybe there's something wrong with my database config? I just installed pgsql on Ubuntu latest, make a db and its owner user also password, and updated the config in setting.py, and then error occurs. Actually I am trying to make it docker image with GitHub actions, but I still cannot runserver locally, and quite confused with the config on proxy as there's no docs

**Comment by doubaniux** [comment]
> > @joseph-cheung there's a comment above :
> > [#92 (comment)](https://github.com/doubaniux/boofilsic/pull/92#discussion_r708789055)
> > just uncomment that line and you should be fine
> > the reason I did not uncomment in the MR is bc I don't know why it's missed in the first place.
> 
> Sorry but I did uncommented that line as you said above... Maybe there's something wrong with my database config? I just installed pgsql on Ubuntu latest, make a db and its owner user also password, and updated the config in setting.py, and then error occurs. Actually I am trying to make it docker image with GitHub actions, but I still cannot runserver locally, and quite confused with the config on proxy as there's no docs

I guess I know the reason, try to comment out the sync related code in `sync.apps` for your first migration

File: boofilsic/settings.py [hunk 3]
```diff
@@ -54,6 +56,7 @@
     'movies.apps.MoviesConfig',
     'music.apps.MusicConfig',
     'games.apps.GamesConfig',
+    # 'sync.apps.SyncConfig',
```
[the remark this answers is not in the document]
Comment (doubaniux) [reply boofilsic/settings.py:59]: Might be some local change unexpectedly committed, well spotted

File: users/models.py [hunk 4]
```diff
@@ -3,18 +3,25 @@
...
 
 
 def report_image_path(instance, filename):
-    return GenerateDateUUIDMediaFilePath(instance, filename, REPORT_MEDIA_PATH_ROOT)
+    return GenerateDateUUIDMediaFilePath(instance, filename, settings.REPORT_MEDIA_PATH_ROOT)
 
 
 class User(AbstractUser):
-    mastodon_id = models.IntegerField(blank=False)
+    if settings.MASTODON_ALLOW_ANY_SITE:
```
[the remark this answers is not in the document]
Comment (doubaniux) [reply users/models.py:17]: I don't get the purpose of username here, could you explain?

File: users/templates/users/register.html [hunk 5]
```diff
@@ -19,15 +19,15 @@
         <img src="{% static 'img/logo.svg' %}" class="logo" alt="boofilsic logo">
 
         <div id="loginButton">
-            <p>欢迎来到NiceDB书影音！</p>
+            <p>欢迎来到{{ site_name }}书影音！</p>
             <p>
-                NiceDB书影音继承了长毛象的用户关系，比如您在里瓣屏蔽了某人，那您将不会在书影音的公共区域看到TA的痕迹。
+                {{ site_name }}书影音继承了长毛象的用户关系，比如您在里瓣屏蔽了某人，那您将不会在书影音的公共区域看到TA的痕迹。
```
Comment (doubaniux) [inline users/templates/users/register.html:24]:
```suggestion
                {{ site_name }}书影音继承了联邦宇宙的用户关系，比如您在联邦宇宙屏蔽了某人，那您将不会在书影音的公共区域看到TA的痕迹。
```

**Comment by doubaniux** [comment]
Is there a reason to use the form `settings.CONSTANTS`? @alphatownsman

File: users/models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (alphatownsman) [reply users/models.py:17]: redefine the username field so it does not have to be unique. in my case, there are user@m.cmx.im and user@alive.bar

**Comment by alphatownsman** [comment]
@joseph-cheung make sure you do these in psql:
```
CREATE DATABASE yourdb ENCODING 'UTF8' LC_COLLATE='en_US.UTF-8' LC_CTYPE='en_US.UTF-8' TEMPLATE template0;
\c yourdb;
CREATE EXTENSION hstore WITH SCHEMA public;
```

and these in project dir: 
```
python3 -m pip -r requirements.txt
python3 manage.py makemigrations users books movies games music sync mastodon management
python3 manage.py migrate users 
python3 manage.py migrate
```

**Comment by alphatownsman** [comment]
@doubaniux not a programmer myself but that's what django doc says: https://docs.djangoproject.com/en/3.2/topics/settings/#using-settings-in-python-code

and that seems to be the only way I can create my own conf file, and use it as DJANGO_SETTINGS_MODULE

**Comment by alphatownsman** [comment]
the default value of MASTODON_CLIENT_SCOPE is r/w/f as it was before, so that this commit won't break NiceDB or other existing running sites (is there any?).

to change this value will require a purge of `django_session`, `mastodon_crosssiteuserinfo` and `mastodon_mastodonapplication` which in theory is safe. `mastodon_mastodonapplication` will have to be repopulated, which is automatic if `MASTODON_ALLOW_ANY_SITE` is true, but otherwise will need some manual setup.

**Comment by alphatownsman** [comment]
@doubaniux consider merging this?

**Comment by joseph-cheung** [comment]
Hello @alphatownsman , I made it docker image with GitHub actions today, but there are a few problems. As you are active on the repo, hopefully that you're kind enough to give me some hint. 

It just gives me 500 error when `STATICFILES_STORAGE = 'django.contrib.staticfiles.storage.ManifestStaticFilesStorage'` is set in settings.py, for all static files are not accessible.
`django.contrib.staticfiles.storage.StaticFilesStorage` just works fine with dynamic pages, but all static files returns 404. 
New to Django, confused, using docker to run Django behind Nginx reverse proxy.

Also I am totally puzzled with the proxy configs, any suggestions?

Now succeeded with docker automatic docker image build with GitHub actions and push to DockerHub, which integrates with pgsql-init on its first run. 
That's to say that we can run it onekey with docker-composer, and it is tested. All problems are about static files.

**Comment by joseph-cheung** [comment]
> Hello @alphatownsman , I made it docker image with GitHub actions today, but there are a few problems. As you are active on the repo, hopefully that you're kind enough to give me some hint.
> 
> It just gives me 500 error when `STATICFILES_STORAGE = 'django.contrib.staticfiles.storage.ManifestStaticFilesStorage'` is set in settings.py, for all static files are not accessible.
> `django.contrib.staticfiles.storage.StaticFilesStorage` just works fine with dynamic pages, but all static files returns 404.
> New to Django, confused, using docker to run Django behind Nginx reverse proxy.
> 
> Also I am totally puzzled with the proxy configs, any suggestions?
> 
> Now succeeded with docker automatic docker image build with GitHub actions and push to DockerHub, which integrates with pgsql-init on its first run.
> That's to say that we can run it onekey with docker-composer, and it is tested. All problems are about static files.

BTW, it all works well with DEBUG=True.
But occurs error with DEBUG=False.
Wired...

**Comment by alphatownsman** [comment]
@joseph-cheung I think if you turn off debug, you have to do `collectstatic` and serve generated files thru nginx.
more in https://docs.djangoproject.com/en/3.2/howto/static-files/deployment/

**Comment by doubaniux** [comment]
@alphatownsman Looks like you've changed your master branch. Could please create a new PR like this one?

**Comment by alphatownsman** [comment]
I closed this PR a few days ago as the next step was not clear, and the differences have been [growing](https://github.com/alphatownsman/boofilsic/commits/neo) since last time you reviewed, so now I'm using `master` to track your code, and `neo` to for mine.

IMHO we can do one of the three pathes:
 - reopen this PR as it was
 - create one big PR from neo to your master, and make features you don't want optional with configuration flags, similar to what I've been trying to do last time, but with more feature and feature flags. it's increasing difficult as more db changes and usage of redis; and it's very hard to keep some feature like caching follower list for faster visibility calculation co-exist with old implementation
 - create multiple small PRs for fixes and features you like to see on NiceDB, it takes time but we can tackle one feature a time as you wish.

I don't have a strong preference, just want to find a proper way to contribute back, let me know which path you prefer.

**Comment by doubaniux** [comment]
@alphatownsman 

> I closed this PR a few days ago as the next step was not clear, and the differences have been [growing](https://github.com/alphatownsman/boofilsic/commits/neo) since last time you reviewed, so now I'm using `master` to track your code, and `neo` to for mine.
> 
> IMHO we can do one of the three pathes:
> 
> * reopen this PR as it was
> * create one big PR from neo to your master, and make features you don't want optional with configuration flags, similar to what I've been trying to do last time, but with more feature and feature flags. it's increasing difficult as more db changes and usage of redis; and it's very hard to keep some feature like caching follower list for faster visibility calculation co-exist with old implementation
> * create multiple small PRs for fixes and features you like to see on NiceDB, it takes time but we can tackle one feature a time as you wish.
> 
> I don't have a strong preference, just want to find a proper way to contribute back, let me know which path you prefer.

Thank you for your contribution so long. If it is still possible to you, I'd like to start to merge features one by one from neodb

**Comment by alphatownsman** [comment]
```
$ git diff --stat master neo | tail -n 1                                 
 192 files changed, 10762 insertions(+), 6141 deletions(-)
```

The change is much larger now, and many features are intertwined, so it might be difficult to cherry pick one by one, but I'll help if that's what you want.

I don't have full list of features, by just to dump what I remembered:
-     Goodreads scraper
-     TMDB scraper
-     GoogleBooks scraper
-     Douban scraper improvements
-     Bandcamp scraper
-     full text search
-     async search external results and one click import
-     Share to public timeline
-     Show ratings from who you follow
-     cache follow/block for performance
-     external sync queue management
-     Export marks
-     List my reviews
-     Delete account
-     3 level of visibility
-     Batch change visibility
-     Collection
-     Fix irregular mastodon domains
-     personal tag view
-     Move account
-     Login via  Friendica, PixelFed
-     Login via Twitter
-     paste image in markdownx
-     link books/movies from different source
-     Import GoodReads shelf and profile
-     wish button in list view
-     Import Douban Review
-     Timeline view

**Comment by Shaokun-X** [comment]
I think I would merge most of them, I'll sort it out, thank you

**Comment by alphatownsman** [comment]
> I think I would merge most of them, I'll sort it out, thank you

Wonderful. I can feature flag those you don’t want. Would you like me to open one big pr?

**Comment by Shaokun-X** [comment]
> > I think I would merge most of them, I'll sort it out, thank you
> 
> Wonderful. I can feature flag those you don’t want. Would you like me to open one big pr?

If every feature has a branch, I think it would be better to merge one by one

**Comment by alphatownsman** [comment]
> If every feature has a branch, I think it would be better to merge one by one

Unfortunately not. I've switched back to single branch mode for quite a while, so most features and bug fixes are intertwined in one branch. IMHO what we can do is one of the following:

 1.   I add feature flags for those functionalities NiceDB do not want, then open one big pull req for you to review and merge
 2.  You list what feature you prefer to have in order, I can recreate branches one by one for you

(1) will be much less labor compare to (2), and make things easier if we want to contribute to each other's code base in the future, hence personally I strongly prefer (1). However, I have huge respect all you've done, so will follow your preference on whichever way you choose to go from here.

**Comment by Shaokun-X** [comment]
@alphatownsman Option one would be cool. Will there be any database migration to the existing data?

**Comment by doubaniux** [comment]
@alphatownsman Option one would be cool. Will there be any migrations to the existing data if all new features would be merged?

**Comment by alphatownsman** [comment]
let's continue in #115

**Pull request closed without merging**